mithril-common/src/chain_observer/pallas_observer.rs

```diff
@@ -78,6 +83,81 @@ impl PallasChainObserver {
         Ok(epoch)
     }
 
+    /// Returns inline datum bytes from the given `Values` instance.
+    fn get_datum_bytes(&self, utxo: &Values) -> Vec<u8> {
+        let bytes = utxo.inline_datum.as_ref().unwrap().1.clone();
+        let bytes = CborWrap(bytes).to_vec();
+        inspect(bytes)
+    }
+
+    /// Returns inline datums from the given `Values` instance.
+    fn inspect_datum(&self, utxo: &Values) -> Datum {
+        let datum = self.get_datum_bytes(utxo);
+
+        inspect(datum)
+    }
+
+    /// Serializes datum to `TxDatum` instance.
+    fn serialize_datum(&self, utxo: &Values) -> TxDatum {
+        let datum = self.inspect_datum(utxo);
+        let serialized = serde_json::to_string(&datum.to_json()).expect("Failed to serialize");
+
+        TxDatum(serialized)
+    }
+
+    /// Maps the given `UTxOByAddress` instance to Datums.
+    fn map_datums(&self, transaction: UTxOByAddress) -> Datums {
+        transaction
+            .utxo
+            .iter()
+            .filter(|(_, utxo)| utxo.inline_datum.is_some())
+            .map(|(_, utxo)| self.serialize_datum(utxo))
+            .collect()
+    }
+
+    /// Returns a vector of `TxDatum` instances.
+    async fn get_utxo_datums(
+        &self,
+        client: &mut NodeClient,
+        address: &ChainAddress,
+    ) -> Result<Datums, ChainObserverError> {
+        let statequery = client.statequery();
+        let utxo = self.get_utxo_by_address(statequery, address).await?;
+
+        Ok(self.map_datums(utxo))
+    }
+
+    /// Fetches the current UTxO by address using the provided `statequery` client.
+    async fn get_utxo_by_address(
+        &self,
+        statequery: &mut Client,
+        address: &ChainAddress,
+    ) -> StdResult<UTxOByAddress> {
+        statequery
+            .acquire(None)
+            .await
+            .map_err(|err| anyhow!(err))
+            .with_context(|| "PallasChainObserver Failed to acquire statequery")?;
+
+        let era = queries_v16::get_current_era(statequery)
+            .await
+            .map_err(|err| anyhow!(err))
+            .with_context(|| "PallasChainObserver Failed to get current era")?;
```

Comment:
Can you fix this typo (which is also in `get_epoch` function)?
```suggestion
            .with_context(|| "PallasChainObserver failed to get current era")?;
```